### ISSUE TITLE
'journalctl --merge' does not interleave entries from different machines

### ISSUE THREAD

**Issue by AndersBlomdell** [systemd/systemd#34169]
### systemd version the issue has been seen with

255

### Used distribution

Fedora Linux 40

### Linux kernel version used

6.10.5-200.fc40.x86_64

### CPU architectures issue was seen on

x86_64

### Component

journalctl

### Expected behaviour you didn't see

`journalctl --merge` returning entries sorted by date

### Unexpected behaviour you saw

`journalctl --merge` returning entries sorted by machine (or more likely by boot-id)

The commit that broke the previous behaviour seems to be (pointed out by a `git bisect`)

```
commit 262299dccbb09f36c8c830dabd6a104469d9852b
Author: Lennart Poettering <lennart@poettering.net>
Date:   Wed Feb 8 11:10:49 2023 +0100

    journal: use boot-id/timestamp info for odering entries
    
``` 

### Steps to reproduce the problem

`journalctl --merge` on a machine with journals from multiple remote sources

**Comment by AndersBlomdell** [comment]
The bisect log should of course have been supplied:
```
git bisect start
# status: waiting for both good and bad commits
# good: [477fdc5afed0457c43d01f3d7ace7209f81d3995] NEWS: finalize
git bisect good 477fdc5afed0457c43d01f3d7ace7209f81d3995
# status: waiting for bad commit, 1 good commit known
# bad: [1e17e48b96bb509754a0a11ea8bd0394965564c6] test: mount ld.so.cache in minimal nspawn container if present
git bisect bad 1e17e48b96bb509754a0a11ea8bd0394965564c6
# bad: [d1f9a39ef49d96d42689a5a1b7a874be1034b3e6] Merge pull request #29853 from YHNdnzj/sleep-automated
git bisect bad d1f9a39ef49d96d42689a5a1b7a874be1034b3e6
# bad: [a130b095133b3f169ba14cdbbe6b3e3fb05686be] Merge pull request #27713 from ddstreet/tpm2_replace_make_primary
git bisect bad a130b095133b3f169ba14cdbbe6b3e3fb05686be
# bad: [7f3bb8f20dcccaceea8b1ee05f0560b81162037b] coredump filter: add mask for 'all' using UINT32_MAX, not UINT64_MAX
git bisect bad 7f3bb8f20dcccaceea8b1ee05f0560b81162037b
# bad: [06d104d58ffa23c958b9b2a2809c61fb25e6f762] bootctl: 'graceful' should ignore EFI variable failures
git bisect bad 06d104d58ffa23c958b9b2a2809c61fb25e6f762
# bad: [cfba58fe79a040386f36aa341c0bbe5bc359faf6] Merge pull request #26651 from yuwata/meson-cleanups
git bisect bad cfba58fe79a040386f36aa341c0bbe5bc359faf6
# good: [e774d3c9f11eb1767202ca4af38f2d3a9238c046] systemctl: warn if units disabled in user scope are still enabled globally
git bisect good e774d3c9f11eb1767202ca4af38f2d3a9238c046
# good: [0a75dca1cc427cbea4db0b5f052da06285d43733] Merge pull request #26579 from poettering/manager-env-clean-up
git bisect good 0a75dca1cc427cbea4db0b5f052da06285d43733
# good: [adee01643d91336f85b8acb82edc23d827edaa7f] Merge pull request #26393 from poettering/mempress
git bisect good adee01643d91336f85b8acb82edc23d827edaa7f
# good: [87e0fd575e8b5a747d25b0bd4cf5a7061c6f6ece] time-util: extend comment a bit
git bisect good 87e0fd575e8b5a747d25b0bd4cf5a7061c6f6ece
# bad: [c63bfd0884cf20e48befbee49d41f667660a8802] systemctl: explicitly cast the constants to uint64_t
git bisect bad c63bfd0884cf20e48befbee49d41f667660a8802
# bad: [16a9ad557de7173c182e9587a9cc0ca146293ec8] Merge pull request #26355 from poettering/journal-no-rtc
git bisect bad 16a9ad557de7173c182e9587a9cc0ca146293ec8
# bad: [262299dccbb09f36c8c830dabd6a104469d9852b] journal: use boot-id/timestamp info for odering entries
git bisect bad 262299dccbb09f36c8c830dabd6a104469d9852b
# good: [34af74946e8853411f18120007ebaca6549b2a52] sd-journal: track newest open journal file per boot ID
git bisect good 34af74946e8853411f18120007ebaca6549b2a52
# first bad commit: [262299dccbb09f36c8c830dabd6a104469d9852b] journal: use boot-id/timestamp info for odering entries
```

**Comment by AndersBlomdell** [comment]
It seems like the later versions skips info as well, previously used version gives a reasonable amount of data

```
# journalctl --version 
systemd 253 (253.17-1.fc38)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP -GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN -IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD +BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=unified
# journalctl --merge --since '2024-08-28 01:00:00' --until '2024-08-28 02:00:00' | wc -l 
5558
```
as opposed to the current version:

```
# ./build/journalctl --version 
systemd 255 (255)
+PAM +AUDIT +SELINUX -APPARMOR +IMA +SMACK +SECCOMP +GCRYPT +GNUTLS +OPENSSL +ACL +BLKID +CURL +ELFUTILS +FIDO2 +IDN2 -IDN +IPTC +KMOD +LIBCRYPTSETUP +LIBFDISK +PCRE2 +PWQUALITY +P11KIT +QRENCODE +TPM2 +BZIP2 +LZ4 +XZ +ZLIB +ZSTD -BPF_FRAMEWORK +XKBCOMMON +UTMP +SYSVINIT default-hierarchy=legacy
# ./build/journalctl --merge --since '2024-08-28 01:00:00' --until '2024-08-28 02:00:00' | wc -l
117
```

**Comment by yuwata** [comment]
This should be a variation of #31516.

**Comment by AndersBlomdell** [comment]
> This should be a variation of #31516.

Maybe, even if I don't find any mention of `--merge` in that bug

If I understand the #31516 report correctly, v255 generates incorrect journals, but v253 can handle them correctly (the are uploaded via jurnal-upload), is a fix for v255 backported yet?

**Comment by yuwata** [comment]
Indeed, the issue #31516 is caused with a buggy journald and RTC batterly less system.
But, actually the offending commit also causes issues in more wider cases, like showing journal generated by journal-upload... Huh, terrible.

**Comment by yuwata** [comment]
> is a fix for v255 backported yet?

I do not think so. But, that's a fix for journald, that is, the side of generating .journal files. Not for showing .journal files, like journalctl. So, even if the fix is backported, your issue is not solved.

**Comment by AndersBlomdell** [comment]
Runnig the tests below (between v253 and main [v256-1716-gbd7b6c213b]) gives surprising results
its simply counting the number of logentries during the last 1000 seconds, both for an open ended
range, and a range that ends at the start of the test.

```
#!/bin/bash

for i in {1..4} ; do

    ROOT=$(dirname $0)
    END=$(date +%s)
    START=$(( ${END} - 1000 ))

    echo
    echo Batch $i
    echo
    
    RANGE="--since @${START}"
    echo ${RANGE}
    echo "v253 $(${ROOT}/build-v253/journalctl --merge ${RANGE} | wc -l)"
    echo "main $(${ROOT}/build/journalctl --merge ${RANGE} | wc -l)"
    
    RANGE="--since @${START} --until @${END}"
    echo ${RANGE}
    echo "v253 $(${ROOT}/build-v253/journalctl --merge ${RANGE} | wc -l)"
    echo "main $(${ROOT}/build/journalctl --merge ${RANGE} | wc -l)"

    RANGE="--since @${START}"
    echo ${RANGE}
    echo "v253 $(${ROOT}/build-v253/journalctl --merge ${RANGE} | wc -l)"
    echo "main $(${ROOT}/build/journalctl --merge ${RANGE} | wc -l)"
 

done
```

The results are:
```
Batch 1

--since @1725030155
v253 8198
main 8205
--since @1725030155 --until @1725031155
v253 8194
main 6951
--since @1725030155
v253 8221
main 8221

Batch 2

--since @1725030159
v253 8192
main 8192
--since @1725030159 --until @1725031159
v253 8184
main 7956
--since @1725030159
v253 8202
main 8208

Batch 3

--since @1725030163
v253 8189
main 8196
--since @1725030163 --until @1725031163
v253 8177
main 5716
--since @1725030163
v253 8215
main 8221

Batch 4

--since @1725030166
v253 8184
main 8184
--since @1725030166 --until @1725031166
v253 8181
main 7244
--since @1725030166
v253 8205
main 8208
```

Where the bounded ranges for `main` seems to be more or less random...

**Comment by AndersBlomdell** [comment]
> Indeed, the issue #31516 is caused with a buggy journald and RTC batterly less system. But, actually the offending commit also causes issues in more wider cases, like showing journal generated by journal-upload... Huh, terrible.

BTW: which is the buggy commit in #31516 ? I really can't deduce it from the issue; is it the same commit as the one my bisect landed at?

**Comment by yuwata** [comment]
> > Indeed, the issue #31516 is caused with a buggy journald and RTC batterly less system. But, actually the offending commit also causes issues in more wider cases, like showing journal generated by journal-upload... Huh, terrible.
> 
> BTW: which is the buggy commit in #31516 ? I really can't deduce it from the issue; is it the same commit as the one my bisect landed at?

The issue #31516 is a combined issue of journald and journalctl, that is, depends on both generator side and reader side. The issue in journald is caused by e5d60d1b3b0c0b229571823c7fdd390a9562982e (v254), and fixed by 3c85d3fda50f71755aa276cfa60807c315bfc04b (v256). The issue in journalctl is caused by 262299dccbb09f36c8c830dabd6a104469d9852b (v254), and is not fixed yet.

**Comment by yuwata** [comment]
Let me confirm, everything fine without --merge?

**Comment by AndersBlomdell** [comment]
> Let me confirm, everything fine without --merge?
Yes!

The culprit seems to be `compare_boot_ids`, unconditionally returning 0 brings the results
back to what is expected. Some fprintf debugging revealed that   `x->newest_machine_id` and  `y->newest_machine_id` are always the same even though `a` and `b` are distinct, so either all my machines have  the same id or `journal_file_find_newest_for_boot_id` is broken.

**Comment by AndersBlomdell** [comment]
All the JournalFiles report the same machine_id, how do I analyze the actual journal-file content?

The machine_id coincides with the machine_id reported when running without --merge`

**Comment by AndersBlomdell** [comment]
Seems like `systemd-journal-remote` inserts the local hosts `machine_id` in the files
it creates, not something I expected :-(

**Comment by AndersBlomdell** [comment]
A small consolation is that the log entries seem to contain the correct `_MACHINE_ID` when doing `journalctl --merge --export`, so a workaround would be to glean the `header.machine_id` from the first entry in thel journal file...

**Comment by AndersBlomdell** [comment]
Unfortunately `systemd-journal-remote` complies with [JOURNAL_FILE_FORMAT](https://systemd.io/JOURNAL_FILE_FORMAT/), which states:

_If a writer finds that the machine_id of a file to write to does not match the machine it is running on it should _immediately_ rotate the file and start a new one._

**Comment by AndersBlomdell** [comment]
> A small consolation is that the log entries seem to contain the correct `_MACHINE_ID` when doing `journalctl --merge --export`, so a workaround would be to glean the `header.machine_id` from the first entry in thel journal file...

Which turned out to be more complex than I had expected, since (AFAICT) we need to iterate over objects in each specific file, we have to create a new iterator that only iterates a file (and not the journal) and then find the proper `_MACHINE_ID` field, convert it to a `sd_id128_t` and then save it in the `file->header`.

The chance that I'll get that right is quite slim, so I'll leave it until Tuesday at the earliest (in the [vain?] hope that somebody more versed in the internals picks up the baton before)

**Comment by AndersBlomdell** [comment]
My current belief is that the machine_id/boot_id ordering is impossible as soon as have logs that spans more than one machine_id (as found  in the DATA/FIELD, not the header).

So my suggestion is to short-circuit the machine_id/boot_id as soon as we have seen more than one   machine_id. The solution I propose is to scan each newly opened file for machine_id's  (and for currently active log-files continue to do so), and stop this scanning as soon as more than one machine-id is seen.

**Comment by yuwata** [comment]
Yeah, I agree with that basically. We need to reconsider the offending commit 262299dccbb09f36c8c830dabd6a104469d9852b.

**Comment by mguentner** [comment]
I can confirm that 262299d / https://github.com/systemd/systemd/pull/26355 is at least in part responsible for this issue:

**Scenario**:
A host receives the logs from several other machines using `systemd-journal-remote` and makes them and its own available through `systemd-journal-gatewayd`, _using_ `--merge`. When queried through `gatewayd`, log entries are randomly (per boot) missing while they are all readable through `journalctl` or `gatewayd`, _without_ `--merge`.

I produced a full revert of https://github.com/systemd/systemd/pull/26355 by doing  `git revert 16a9ad557de7173c182e9587a9cc0ca146293ec8 -m 1` against v256.4.
You can find the patch [here](https://gist.github.com/mguentner/5843b0f433a71f85cdf23299b49065db). systemd compiles cleanly, however tests, README and TODO might not make sense anymore.
Since I needed to resolve quite a lot of conflicts manually, it may contain new bugs ;).

I then applied the patch to build a new system. Even after many reboots (new boot ids), I am not able to reproduce the issue anymore.

**Comment by AndersBlomdell** [comment]
> This is probably the same as #8979/#19744?
Nope, since those predates the commit that broke the ordering




> My current belief is that the machine_id/boot_id ordering is impossible as soon as have logs that spans more than one machine_id (as found in the DATA/FIELD, not the header).
> 
> So my suggestion is to short-circuit the machine_id/boot_id as soon as we have seen more than one machine_id. The solution I propose is to scan each newly opened file for machine_id's (and for currently active log-files continue to do so), and stop this scanning as soon as more than one machine-id is seen.
This is still what I think is needed.

**Comment by mator** [comment]
Any updates? Can we revert #26355 or the bug will stay in all versions after 255 ?

**Comment by keszybz** [comment]
Sorry, we didn't get around to this at all this release.

**Comment by msekletar** [comment]
I retested the scenario with today's main and it seems to work for me (i.e. I can obtain all entries stored in two journals from two different machines via gatewayd with --merge and --directory + entries are interleaved properly). Maybe the issue previously was in how journals were generated and now with new enough journald, the generation (I've generated journals with systemd-v259) is fixed and hence merging works correctly as well.

@AndersBlomdell can you double check the current state and close the issue if appropriate.

**Comment by mrc0mmand** [comment]
I played around with this locally and I can still reproduce the issue. As was mentioned earlier, the root cause is that `compare_boot_ids()` skips the tail entry realtime timestamp ordering only if the machine IDs of the two journal files differ, but journals imported via systemd-journal-remote have the machine ID in the journal header set to the host's machine ID, defeating this check. This then causes journalctl to group and order entries with specific boot IDs one after another instead of properly interleaving them.

I'm not sure if we want to revert https://github.com/systemd/systemd/commit/262299dccbb09f36c8c830dabd6a104469d9852b. The proposed solution of scanning the journal file and flagging it if it contains an entry with a different `_MACHINE_ID=` than what's in the journal's header might work, but would be, IMO, quite expensive, at least in some cases.

After thinking about this a bit, one potential way out could be updating the journal file `newest_machine_id` field based on the `_MACHINE_ID=` field value of the latest tail entry. We already call `journal_file_find_newest_for_boot_id()` from `compare_boot_ids()`, which in turn calls `journal_file_read_tail_timestamp()`. In this function we always update the `newest_machine_id` field with a machine ID from the file header:

https://github.com/systemd/systemd/blob/f807a6f26d150d9e8138ef59d2ff2c9c7e860d39/src/libsystemd/sd-journal/sd-journal.c#L2694-L2696

Given we already have the tail entry at hand, we can scan it for the `_MACHINE_ID=` field and use that value instead of what's in the header to update `newset_machine_id`. This would then cause `compare_boot_ids()` to actually return _before_ doing the tail entry realtime timestamp check, and `compare_locations()` would instead fall trough to a current entry realtime check, and we _should_ end up with a proper interleaving. There would still be _some_ overhead as we need to iterate through the entry fields to find the `_MACHINE_ID=` field, but it shouldn't be _that_ bad.

I'm not, yet, sure if using the tail entry's machine ID instead of the one from the journal file header might break something else - if it does, we could probably set a flag on the journal file if the header has a different machine ID than the tail entry.

**Comment by AndersBlomdell** [comment]
First of all: Thank you for looking into this!, I'm swamped with other stuff ATM

> I played around with this locally and I can still reproduce the issue. As was mentioned earlier, the root cause is that `compare_boot_ids()` skips the tail entry realtime timestamp ordering only if the machine IDs of the two journal files differ, but journals imported via systemd-journal-remote have the machine ID in the journal header set to the host's machine ID, defeating this check. This then causes journalctl to group and order entries with specific boot IDs one after another instead of properly interleaving them.
> 
> I'm not sure if we want to revert [262299d](https://github.com/systemd/systemd/commit/262299dccbb09f36c8c830dabd6a104469d9852b). The proposed solution of scanning the journal file and flagging it if it contains an entry with a different `_MACHINE_ID=` than what's in the journal's header might work, but would be, IMO, quite expensive, at least in some cases.
> 
> After thinking about this a bit, one potential way out could be updating the journal file `newest_machine_id` field based on the `_MACHINE_ID=` field value of the latest tail entry. We already call `journal_file_find_newest_for_boot_id()` from `compare_boot_ids()`, which in turn calls `journal_file_read_tail_timestamp()`. In this function we always update the `newest_machine_id` field with a machine ID from the file header:
> 
> https://github.com/systemd/systemd/blob/f807a6f26d150d9e8138ef59d2ff2c9c7e860d39/src/libsystemd/sd-journal/sd-journal.c#L2694-L2696
> 
> Given we already have the tail entry at hand, we can scan it for the `_MACHINE_ID=` field and use that value instead of what's in the header to update `newset_machine_id`. This would then cause `compare_boot_ids()` to actually return _before_ doing the tail entry realtime timestamp check, and `compare_locations()` would instead fall trough to a current entry realtime check, and we _should_ end up with a proper interleaving. There would still be _some_ overhead as we need to iterate through the entry fields to find the `_MACHINE_ID=` field, but it shouldn't be _that_ bad.
> 
> I'm not, yet, sure if using the tail entry's machine ID instead of the one from the journal file header might break something else - if it does, we could probably set a flag on the journal file if the header has a different machine ID than the tail entry.

It might not work if `--split-mode=none` or if output is directed to a single file (which I have never tried)

**Comment by mrc0mmand** [comment]
> It might not work if `--split-mode=none` or if output is directed to a single file (which I have never tried)

I _think_ that should be fine, because in that case the entries should be already properly interleaved, as they should get the seqnum based on the order they were received in. And since all entries would be in a single file with a single seqnum_id, we'd never fall through to `compare_boot_ids()`.

**Comment by AndersBlomdell** [comment]
> > It might not work if `--split-mode=none` or if output is directed to a single file (which I have never tried)
> 
> I _think_ that should be fine, because in that case the entries should be already properly interleaved, as they should get the seqnum based on the order they were received in. And since all entries would be in a single file with a single seqnum_id, we'd never fall through to `compare_boot_ids()`.

What happens if the last entry is uploaded from the local host (yes, it's a dubious setup, duplicating entries)

**Comment by mrc0mmand** [comment]
> What happens if the last entry is uploaded from the local host (yes, it's a dubious setup, duplicating entries)

Nothing would change - the entry would still have the same seqnum_id as the rest of the entries in that file, so we'd still go through a seqnum comparison instead of `compare_boot_ids()`.
